# Chapter 5
**Sequnces: Strings, Lists, and Files

5.1 The String data types

In [3]:
# String Types example

str1 = "Hello"
str2 = 'spam'
print(str1, str2)

Hello spam


In [4]:
type(str1)

str

In [5]:
type(str2)

str

In [6]:
# Indexing Example

greet = "Hello Bob"

In [7]:
greet[0]

'H'

In [8]:
print(greet[0], greet[2], greet[4])

H l o


In [9]:
x = 8

In [10]:
print(greet[x-2])

B


In [11]:
greet[-1]

'b'

In [12]:
greet[-3]

'B'

In [13]:
# Slicing example

In [14]:
greet[0:3]

'Hel'

In [15]:
greet[5:9]

' Bob'

In [16]:
greet[:5]

'Hello'

In [17]:
greet[5:]

' Bob'

In [18]:
greet[:]

'Hello Bob'

In [19]:
# Examples of various string operations

In [20]:
"spam" + "eggs"

'spameggs'

In [22]:
"Spam" + "And" + "Eggs"

'SpamAndEggs'

### 

In [23]:
3 * "spam"

'spamspamspam'

In [24]:
"spam" * 5

'spamspamspamspamspam'

In [25]:
(3 * "spam") + ("eggs" * 5)

'spamspamspameggseggseggseggseggs'

In [26]:
len("spam")

4

In [27]:
len("SpamAndEggs")

11

In [29]:
for ch in "Spam!":
    print(ch, end=" ")

S p a m ! 

5.2 Simple String Processing

In [32]:
# username.py
#   Simple string processing prgram to generate usernames.

def main():
    print("This program generates computer usernames. \n")
    
    # get users's first and last names
    first = input("Please enter your first name (all lower cases): ")
    last = input("Please eneter your last name (all lower cases) ")
    
    # concatenate first inital with 7 chars of the last name.
    uname = first[0] + last[:7]
    
    # output the username
    print("Your username is:", uname)
    
main()

This program generates computer usernames. 

Please enter your first name (all lower cases): zaphod
Please eneter your last name (all lower cases) beeblebrox
Your username is: zbeebleb


In [33]:
# month.py
# A Program to print the abbreviation of a month, given it's number

def main():
    # months is used as a lookup table
    months = "JanFebMarAprMayJunJulAugSepOctNovDec"
    
    n = int(input("Enter a month number (1-12): "))
    
    # compute starting position of month n in months
    pos = (n-1) * 3
    
    # grab the appropriate slice from months
    monthAbbrev = months[pos:pos+3]
    
    # print the result
    print("The month abbreviation is", monthAbbrev + ".")
    
main()

Enter a month number (1-12): 1
The month abbreviation is Jan.


5.3 Lists as sequences

In [35]:
# Operations for lists example

In [36]:
[1,2] + [3,4]

[1, 2, 3, 4]

In [37]:
[1,2]*3

[1, 2, 1, 2, 1, 2]

In [38]:
grades = ["A","B","C","D","F"]

In [39]:
grades[0]

'A'

In [40]:
grades[2:4]

['C', 'D']

In [41]:
len(grades)

5

In [42]:
myList = [1, "Spam", 4, "U"]

In [44]:
# month2.py
#   A program to print the month abbreviation, given its number.

def main():
    
    # months is a list used as a look up table
    months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
    
    n = int(input("Enter a month number (1-12): "))
    
    print("The month abbreviation is", months[n-1] + ".")
    
main()

Enter a month number (1-12): 1
The month abbreviation is Jan.


In [45]:
# Lists are mutable, whilst strings are not example

In [46]:
myList = [34, 36, 15, 10]

In [47]:
myList[2]

15

In [49]:
myList[2] = 0

In [50]:
myList

[34, 36, 0, 10]

In [51]:
myString = "Hello World"

In [52]:
myString[2]

'l'

In [53]:
myString[2] = "z"

TypeError: 'str' object does not support item assignment

5.4 String Representation and Message Encoding

In [55]:
ord("a")

97

In [56]:
ord("A")

65

In [57]:
chr(97)

'a'

In [58]:
chr(90)

'Z'

In [62]:
# test2numbers.py
#    A Program to convert a textual message into a sequqnce of numbers utilising the underyling Unicode encoding.

def main():
    print("This program converts a textual message into a sequence")
    print(" of numbers representing the Unicode encoding of the message. \n")
    
    # Get the message to encode
    message = input("Please enter the message to encode: ")
    
    print("\nHere are the Unicode codes: ")
    
    # Loop through the message and print out the Unicode values
    for ch in message:
        print(ord(ch),end=" " )
        
    print() # blank line before prompt
    
main()

This program converts a textual message into a sequence
 of numbers representing the Unicode encoding of the message. 

Please enter the message to encode: Sourpuss!

Here are the Unicode codes: 
83 111 117 114 112 117 115 115 33 


5.5 String Methods

In [64]:
# Split strings example

In [65]:
myString = "Hello, string methods!"

In [66]:
myString.split()

['Hello,', 'string', 'methods!']

In [1]:
"32, 24, 25, 57".split(",")

['32', ' 24', ' 25', ' 57']

In [2]:
coords = input("Enter the point of coordinates (x,y): ").split(",")

# Enter the point coordinates (x,y)

coords

Enter the point of coordinates (x,y): 3.4, 6.25


['3.4', ' 6.25']

In [4]:
coords[0]

'3.4'

In [5]:
coords[1]

' 6.25'

In [9]:
coords = input("Enter the point coordinates (x,y): ").split(",")

x,y = flat(coords[0], flat(coords[1]))

Enter the point coordinates (x,y): 3, 6


NameError: name 'flat' is not defined

In [14]:
# numbers2text.py
#   A program to convert a sequence of Unicode numbers into a 
#      a string of text.

def main():
    print("This program converts a sequence of Unicode numbers into")
    print("a string of text")
    
    # Get the message to encode
    inString = input("Please enter the Unicode-encoded message: ")
    
    # Loop through each substring and build Unicode message
    message = " "
    for numStr in inString.split():
        codeNum = int(numStr)            # converts digits to a number
        message = message + chr(codeNum)  # concatentate character to message
        
    print("\nThe decoded message is:", message)
    
main()

This program converts a sequence of Unicode numbers into
a string of text
Please enter the Unicode-encoded message: 83 116 114 105 110 103 115 32 97 114 101 32 70 117 110 33

The decoded message is:  Strings are Fun!


In [16]:
 s = "hello, I came here for an argument"
s.capitalize()

'Hello, i came here for an argument'

In [17]:
s.title()

'Hello, I Came Here For An Argument'

In [18]:
s.lower()

'hello, i came here for an argument'

In [19]:
s.upper()

'HELLO, I CAME HERE FOR AN ARGUMENT'

In [20]:
s.replace("I", "you")

'hello, you came here for an argument'

In [21]:
s.center(30)

'hello, I came here for an argument'

In [22]:
s.center(50)

'        hello, I came here for an argument        '

In [23]:
s.count("e")

5

In [24]:
s.find(",")

5

In [27]:
" ".join(["Number", "one", "the", "Larch"])

'Number one the Larch'

In [28]:
"spam".join(["Number", "one", "the", "Larch"])

'NumberspamonespamthespamLarch'

In [30]:
#5.6 Lists have methods, Too

# This programs starts with an empty list ([])
#   and each number from 1 to 100 and squared and appended to the list
squares = []
for x in range(1,101):
        squares.append(x*x)

In [35]:
# numbers2text.py
#   A program to convert a sequence of Unicode numbers into a 
#      a string of text.

def main():
    print("This program converts a sequence of Unicode numbers into")
    print("a string of text")
    
    # Get the message to encode
    inString = input("Please enter the Unicode-encoded message: ")
    
    # Loop through each substring and build Unicode message
    chars = []
    for numStr in inString.split():
        codeNum = int(numStr)            # converts digits to a number
        chars.append(chr(codeNum))         # accumulate new character
        
    message = "".join(chars)
    print("\nThe decoded message is:", message)
    
main()

# In this code, we call the characters by appending them to a list called chars.
# The final message is obtained by joining these characters together using 
# an empty string as an operator.

This program converts a sequence of Unicode numbers into
a string of text
Please enter the Unicode-encoded message: 83 116 114 105 110 103 115 32 97 114 101 32 70 117 110 33

The decoded message is: Strings are Fun!


5.8 Input/Output as String Manipulation

5.8.1 Example application: Date Converstion

In [2]:
# dataconvert.py
#   Converts a date in form "mm/dd/yyyy" to month day, year"
#   American format

def main():
    # get the date
    dateStr = input("Enter a data (mm/dd/yyyy): ")
    
    # split into components (splits at the "/" of the above)
    monthStr, dayStr, yearStr = dateStr.split("/")
    
    # convert monthStr to the month name
    months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
    monthStr = months[int(monthStr)-1]
    
    # Output result in month day, year format
    print("The converted date is:", monthStr, dayStr+",", yearStr)

main()

Enter a data (mm/dd/yyyy): 05/24/2020
The converted date is: May 24, 2020


In [3]:
# dataconvert.py
#   Converts a date in form "dd/mm/yyyy" to month day, year"
#   Australian format

def main():
    # get the date
    dateStr = input("Enter a data (dd/mm/yyyy): ")
    
    # split into components (splits at the "/" of the above)
    dayStr, monthStr, yearStr = dateStr.split("/")
    
    # convert monthStr to the month name
    months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
    monthStr = months[int(monthStr)-1]
    
    # Output result in month day, year format
    print("The converted date is:", monthStr, dayStr+",", yearStr)

main()

Enter a data (dd/mm/yyyy): 21/09/1995
The converted date is: September 21, 1995


## Notes

Often it is also necessary to turn a number into a stirng. In Python,
most data types can be converted into strings using the str function.

By turning value into a string, we can use string concatenation to put a period at the end of a sentence. If we didn't first turn value into a string, Python would interpret the + as a numerical operation and produce an error, because "." is not a number.

In [4]:
str(500)

'500'

In [5]:
value = 3.14

In [6]:
str(value)

'3.14'

In [7]:
print("The value is", str(value) + ".")

The value is 3.14.


5.8.3 Better change Counter

In [7]:
# change2.py
#    A program to calculate the value of some change in dollars
#    This version represents the total cash in cents.

def main():
    print("Change Counter/n")
    print("Please enter the count of each coin type.")
    quarters = int(input("Quarters: "))
    dimes = int(input("Dimes: "))
    nickels = int(input("Nickles: "))
    pennies = int(input("Pennies: "))
    
    total = quarters * 25 + dimes * 10 + nickels * 5 + pennies
    
    print("The total value of your change is ${}.{}".format(total//100, total%100))
    
main()

Change Counter/n
Please enter the count of each coin type.
Quarters: 10
Dimes: 10
Nickles: 10
Pennies: 10
The total value of your change is $4.10


## 5.9 File Processing

### 5.9.1 Multi-line Strings

Conceptually, a file is a sequence of data that is stored in secondary memory (usually on a disk drive).

### 5.9.2 File Processing

All languages share certain underlying file-manipulation concepts. 

First, we need some way to associate a file on disk with an object in a program. This process is called opening a file.

Second, we need a set of operations that can manipulate the file object. At the very least, this includes operations that allow us to read the information from a file and write new information to a file.

### 5.9.3 Example program: Batch Usernames

In batch processing, program input and output is done through files.

Our new program is designed to process a file of names. Each line of the input file will contain the first and last names of a new user seperated by one or more spaces. The program will produce an output file containing a line for each generated use.

See below 5.9.3 example

### 5.9.4 File dialog (Optional)

In [13]:
# 5.9.2 An example prgoram that prints the contents of a file to the screen
# using the read operation.

# printfile.py
#    Prints a file to the screen.

def main():
    fname = input("Enter filename: ")
    infile = open(fname,"r")
    data = infile.read()
    print(data)
    
main()

# The program first prompts the user for a file name and then opens
# the file for reading through variable infile.

Enter filename: 1


FileNotFoundError: [Errno 2] No such file or directory: '1'

In [18]:
# userfile.py
#    Program to create a file of usernames in batch mode.

def main():
    print("This program creates a file of usernames from a")
    print("file of names")
    
    # get the file names
    infileName = input("What file are the names in? ")
    outfileName = input("What file should the usernames go in ")
    
    # open the files
    infile = open(infileName, "r")
    outfile = open(outfileName, "w")

    # process each line of the input file for line in infile:
    for line in infile:
        # get the first and last names from line
        first, last = line.split()
        # create the username
        uname = (first[0]+last[:7]).lower()
        # write it to the output file
        print(uname, file=outfile)
    
    # close both files
    infile.close()
    outfile.close()
    
    print("Usernames have been written to", outfileName)
    
main()

This program creates a file of usernames from a
file of names
What file are the names in? 1
What file should the usernames go in jj


FileNotFoundError: [Errno 2] No such file or directory: '1'

## Chapter summary

This chapter has covered important elements of the Python string, list and file objects. Here is a summary of highlights.

- Strings are sequences of characters. String literals can be delimited with either single or double quotes.

- Strings and lists can be manipulated with the built-in sequence operations for concatenation (+), repition (*), indexing ([]), slicing ([:]), and length (len()). A for loop can be used to iterate through the characters of a string, items in a list, or lines of a file.

- One way of converting numberic information into string information is to use a string or a list as a lookup table.

- Lists are more general than strings.

  - Strings are always sequences of characters, whereas lists can contain values of any type.
  - Lists are mutable, which means that items in a list can be modified by assigning new values.
  
- Strings are represented in the computer as numeric codes. ASCII and Unicode are compatible standards that are used for specifiying the correspondance between characters and the underlying codes. Python provides the ord and chr functions for translating between Unicode codes and characters.

- Python string and list objects include many useful built-in methods for string and list processing.

- The process of encoding data to keep it private is called encryption. There are two different kinds of encryption systems: private and and public key.

- Program input and output oftem involve string processing. Python provides numerous operators for converting back and forth between numbers and strings. The string formatting methods (format) is particularly useful for producing nucely formatted output.

- Text files are multi-line strings stored in secondary memory. A text file may be opened for readning or writing. When opened for writing, the exisiting contects of the file are erased. Python provides three file-reading methods: read(), readline(), and readlines(). It is also possible to iterate through the lines of a file with a for loop. Data is written to a file using the print function. When processing is finished, a file should be closed.

## 5.11 Exercises

### Review Questions
#### True/ False

1. A Python string literal is always enclosed in double quotes.

False.

A string is a data type for representing a sequence of characters (text). Also, all strings are enclosed in quotations (Single or double) on both sides of the sentence/word.

Example you can use:

"testing" = 'testing'

2. The last character of a string is s is at position len(s)-1.

False. WRONG

Last characters of strings are len(s)+1 as the numerica positions in Python exclude the last number.

#### True.
##### The len function provides the length of the string. But because indexses start at 0, the last position (index) is the length of the string minus one.

3. A string always contains a single line of text.

False.

A string is a data type for representing a sequence of characters. This sequence can be of any length.In fact, it is also possible to have an empty string, i.e. "".

4. In Python "4" + "5" is "45".

True.

As they are in string formats. They will equal the value of "45".

This is not the case of using the additional operator on integers. Since the literals are enclosed in double quotations, we know that we were dealing with values of the data type string. The operator + thus indicates the string oepration, concatenation. 

5. Python lists are mutable, but strings are not.

True.

Lists are more general than strings.

- Strings are always sequences of characters, whereas lists can contain values of any type.

- Lists are mutable, which means that items in a list can be modified by assigning new values.

6. ASCII is a standard for representing characters using numeric codes.

True.

ASCII and the Unicode are compatible standards that are used for specifying the correspondance between characters and the underlying codes. Python provides the ord and chr functions for translating between Unicodes codes and characters.

To avoid this sort of problem, computer systems today use industry standard ecnodings. One important standard is called ASCII (American Standard Code for Information Interchange). ASCII uses the number 0 through 127 to represent characters typically found on an (American) computer keyboard, as well as certain special values known as control codds that are used to coordinate the sending and recieving of information.

7. The split method breaks a string into a list of substrings, and join does the opposite.

True. WRONG.


The split method breaks a string into a list of substring while the join method joins all items into a string, using the hash character as a seperator.

#### False.
##### Split method splits a string into a list of substrings. By default, it will split the string whenever a space occurs.

8. A substitution cipher is a good way to keep sensitive information secure.

True. WRONG.

In a substituion cipher, any character of plain text from the given fixed set of characters is substituted by some other character from the same set depending on the key.

#### False.
##### Our simple  encoding/decoding programs use a very weak form of encryption known as a substitution cipher. Since each letter is always encoded by the same symbol, a code breaker could use statistical information about the frequency of various letters and some simple trial and error testing to discover the original message.


9. The add method can be used to add an item to the end of a list.

False.

Append methods adds an item to the end of a list. This is often used to build a list an item at a time.

10. The process of associating a file with an object in a program is called "reading" the file.

True. WRONG.

#### False.
##### We need some way to associate a file on disk with an object in a program. This process is called opening a file. Once a file has been opened, its contents can be accessing through the associated file object.

## 6/10
## 60% - C

### Multiple Choice

1. Accesing a single character out of a string is called: 

D. Indexing

a. Slicing

Slicing is extracting a subsequence of a string, list, or other sequence object.

b. Concatenation

String concatenation means add strings together. Use the + character to add a variable to another variable.

c. Assignment

The process of giving a value to a variable.

d. Indexing

Selecting a single item from a sequence based on its relative position in the sequence.

2. Which of the following is the same as s[0:-1]? 
 
c.s[0:len(s)-1]

Example s = (0,1,2,3,4,5,6,7,8,9,10)

s[0:-1] 

(1,2,3,4,5,6,7,8,9)

a. s[-1]

10

b. s[:]

(1,2,3,4,5,6,7,8,9,10)

c. s[:len(s)-1]

(1,2,3,4,5,6,7,8,9)

d. s[0:len(s)]

(1,2,3,4,5,6,7,8,9,10)

3. What function gives the Unicode value of a character? 

C. chr

WRONG

a. ord
Python provides a couple of built-in functions that allow us to switch back and forth between characters and the numberic values used to represent them in strings. The ord function returns the numeric ("ordinal") code of a sungle character string, while chr goes the other direction.

a. ord

The ord function returns the number representing the unicode code of a specificed character.

b. ASCII

American Standard Code for Information Interchange. A standard for encoding text where each character is represented by a numnber 0-127)

c. chr

The chr() method returns a string representing a charcater whose Unicode code point is an intger.

d. eval

eval is a built-in function used in python, eval function pasrses the expression argument and evaluates it as a python expression. In simple words, the eval function evaluates the "String" like a python epxression and returns the result as an integer.

4. Which of the following can not be used to convert a string of digits into a number?

D. Eval

a. int
The int method returns an integer object from an number or string.

b. float
The float method returns a floating pint number from a number or a stirng.

c. str

The str function converts values to a string form. 

d. eval

eval is a built-in function used in python, eval function pasrses the expression argument and evaluates it as a python expression. In simple words, the eval function evaluates the "String" like a python epxression and returns the result as an intger.

5. A successor to ASCII that includes characters from (nearly) all written languages is

C. Unicode

a. TELLI

b. ASCII++

c. Unicode

An alternative to ASCII that encodes characters from virtually all of the world's written languages. Unicode is disgned to by ASCII-compatible.

d. ISO

6. Which string method converts all the characters of a string to upper case?

D. upper

a. capitalize

b. capwords

c. uppercase

d. upper

Upper is a built-in method used for string handling. The upper() methods returns the uppercased string from the givein string. It converts all lowervace character to uppercase.

7. The string "slots" that are filled in by the format method are marked by:

The format method returns the formatted string.

<template-string>.format(<values>)

D. {}.

Curly braces inside the template-string mark "slots" inwhich whcich the provided values are inserted.

a. %

b. $

c. []

d. {}


8. Which of the following is not a file reading method in Python?

D. Readlines

WRONG

c. readall

Python provides three related operations for reading information from a file [read,readline, readlines.]

a. read

Returns the entire remaining contents of the file as a single (potentially large, multi-line) string.

b. readline

Returns the next line of the file. That is, all text up to and including the next newline character.

c. readall

d. readlines

Returns a list of the remaining lines in the file. Each list item is a single line including the newline character at the end.

9. The term for a program that does its input and output with files is

a. file-orientated

WRONG

c.Batch

a. file-orientated

b. multi-line

c. batch

d. lame

10. Before reading or writing to a file, a file object must be created via

a. Open

a. open

b. create

c. File

d. Folder

## 7/10
## 70% - B

In [1]:
s = (1,2,3,4,5,6,7,8,9,10)

In [2]:
s[0:-1]

(1, 2, 3, 4, 5, 6, 7, 8, 9)

In [3]:
s[-1]

10

In [4]:
s[:]

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

In [5]:
s[:len(s)-1]

(1, 2, 3, 4, 5, 6, 7, 8, 9)

In [6]:
s[0:len(s)]

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

### Discussion

1. 
Given the initial statements:
    
    s1 = "spam"
    s2 = "ni!"
    
Show the result of evaluating each of the following expressions:
    
    a.
    "The knights who say, " + s2
    'the knights who say, ni!'
    
    b.
    3 * s1 + 2 * s2
    'spamspamspamni!ni!'
    
    c.
    s1[1]
    'p'
    
    d.
    s1[1:3]
    'pa'
    
    e.
    s2[2] + s2[:2]
    '!ni'
    
    f.
    s1+s2[-1]
    'spam!'
    
    g.
    s1.upper()
    'SPAM'
    
    h.
    s2.upper().ljust(4) * 2
    'NI! NI! '
    
    Space at the 3rd count

In [3]:
s1 = "spam"
s2 = "ni!"

In [4]:
"the knights who say, " + s2

'the knights who say, ni!'

In [5]:
3 * s1 + 2 * s2

'spamspamspamni!ni!'

In [6]:
s1[1]

'p'

In [7]:
s1[1:3]

'pa'

In [8]:
s2[2]+s2[:2]

'!ni'

In [9]:
s1+s2[-1]

'spam!'

In [10]:
s1.upper()

'SPAM'

In [11]:
s2.upper().ljust(4)*2

'NI! NI! '

In [14]:
s2.ljust(4)*2

'ni! ni! '

In [19]:
s2.ljust(2)*2

'ni!ni!'

In [16]:
s2*2

'ni!ni!'

2.
Given the same inital staements as in the previous problem, show a Python expression that could construct each of the following results by performing string operations on s1 and s2.

a.
"NI"


b.
"ni!spamni!"


c.
"SPAM NI!  SPAM NI!  SPAM NI!"


d.
"spam"


e.
["sp", "m"]


f.
"spm"



In [23]:
s2[:2].upper()

'NI'

In [24]:
s2+s1+s2

'ni!spamni!'

In [35]:
(s1.capitalize().ljust(6) + s2.capitalize()).ljust(4) * 3

'Spam  Ni!Spam  Ni!Spam  Ni!'

In [38]:
s1[:4]

'spam'

In [54]:
s1.split("a")

['sp', 'm']

In [48]:
s1[:2]

'sp'

In [51]:
s1[3:]

'm'

In [55]:
s1[:2]+s1[3:]

'spm'

3.
Show the output that would be generated by each of the following program fragments:

a. for ch in "aardvark"
       print(ch)
    
   a
   a
   r
   d
   v
   a
   r
   k
     
b. for w in "Now is the swing of our discontent..."split():
       print(w)
    
   Now
   is
   the
   swing
   of
   our
   discontent...


c. for w in "Mississippi".split("i"):
       print(w, end=" ")
    
   m ss ss pp

d. msg = " "
   for s in "secret".split("e"):
       msg = msg + s
   print(msg)
   
   scrt

   for the above Msg = a blank space, consequently beggining the print  state with a space.

e. msg = " "
   for ch in "secret":
       msg = msg + chr(ord(ch)+1)
   print(msg)
   
   tfdsfu
   
   The fragment prints a string by taking the Unicode value of each letter,adding 1 to it, and then evaluating the in to return the corresponding Unicode character, thus altering the unicode of each character in "secret" by +1, by concatenating the result in msg each time the loop iterates.
   
   s = 115 + 1 = 116 = t
   e = 101 + 1 = 102 = f
   c = 99 + 1 = 100 = d
   r = 114 + 1 = 115 = s
   e = 101 + 1 = 102 = f
   t = 116 + 1 = 117 = u

In [1]:
for ch in "aardvark":
    print(ch)

a
a
r
d
v
a
r
k


In [5]:
for w in "Now is the winter of our discontect...".split():
    print(w)

Now
is
the
winter
of
our
discontect...


In [7]:
for w in "Mississippi".split("i"):
    print(w, end = " ")

M ss ss pp  

In [13]:
msg = " "
for s in "secret".split("e"):
    msg = msg + s
print(msg)

 scrt


In [10]:
msg = " "
for ch in "secret":
    msg = msg + chr(ord(ch)+1)
print(msg)

 tfdsfu


In [16]:
msg = " "
for ch in "secret":
    msg = msg + chr(ord(ch)+0)
print(msg)

 secret


4.
Show the string that would result from each of the following string formatting operations. If the operation is not legal, explain why.

a.
"Looks like {1} and {0} for breakfast".format("eggs", "spam")

'Looks like spam and eggs for breakfast'

b.
"There is {0} {1} {2} {3}".format(1, "spam", 4, "you")

'There is 1 spam 4 you'

c.
"Hello {0}".format("Susan", "Computewell")

'Hello Susan'

d.
"{0:0.2f} {0:02f}".format(2.3, 2.3468)

'2.30 2.300000'

e.
"{7.5f} {7.5f}".format(2.3, 2.3468)

Error occured.

The reason the error occured is because there is no tupple at index 7.5.

f.
"Time left {0:02}:{1:05.2f}".format(1, 37.374)

'Time left 01:37.37'

g.
"{1:3}".format("14")

Error occured.

The reason the error occured is because we only have established a tuple at index 0.

In [17]:
"Looks like {1} and {0} for breakfast".format("eggs","spam")

'Looks like spam and eggs for breakfast'

In [19]:
"There is {0} {1} {2} {3}".format(1, "spam", 4, "you")

'There is 1 spam 4 you'

In [20]:
"Hello {0}".format("Susan", "Computewell")

'Hello Susan'

In [21]:
"{0:0.2f} {0:02f}".format(2.3, 2.3468)

'2.30 2.300000'

In [28]:
"{7.5f} {7.5f}".format(2.3, 2.3468)

IndexError: tuple index out of range

In [23]:
"Time left {0:02}:{1:05.2f}".format(1, 37.374)

'Time left 01:37.37'

In [26]:
"{1:3}".format("14")

IndexError: tuple index out of range

5.
Explain why public key encryption is more useful for securing communications on the Internet than private (shared) key encryption.

In a private key (also called shared key) system, the same key is used for encrypting and decrypting messages. All parties that wish to communicate need to know the key, but it must be kept secret from the outside world. The is the usually system people think of when considering secret codes.

In public key systems, there are seperate but related keys for encrypting and decrypting. Knowing the encryption key does not allow you to decrypt messages or discover the decrypting key. In a public key system, the encrypting key can be made pubblicly available, while the decryption key is kept private. Anyone can safely send a message using the pubblic key for encryption. Only the party holding the decryption key will be able to decipher it. For exmaple, a secure website can send your web browser its public key, and the browser can use it to encode your credit card information before sending it to the Internet. Then only the company that is requesting the information will be able to decrypt and read it using the proper private key.

Private key's are more secure than public keys however private keys run the risk of losing encryption and decryption keys which is highly dangerous.

Public keys however can only be decrypted with another key that is specific to the recipient.

### Programming exercises

1.
The example code files for Chapter 5 include a date conversion program dateconvert2.py. This program could be simplified with string formatting. Modify the program to used the string format method for its output.

In [30]:
# dataconvert2.py
#   Converts a date in form "dd/mm/yyyy" to month day, year"
#   Australian format

def main():
    # get the date
    dateStr = input("Enter a data (dd/mm/yyyy): ")
    
    # split into components (splits at the "/" of the above)
    dayStr, monthStr, yearStr = dateStr.split("/")
    
    # convert monthStr to the month name
    months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
    monthStr = months[int(monthStr)-1]
    
    # Output result in month day, year format
    print("The converted date is:", monthStr, dayStr+",", yearStr)

main()

Enter a data (dd/mm/yyyy): 22/09/1995
The converted date is: September 22, 1995


In [7]:
def main():
    # get the day month and year as numbers
    day = int(input("Enter the day number: "))
    month = int(input("Enter the month number: "))
    year = int(input("Enter the year: "))

    # Add components
    date1 = str(month)+"/"+str(day)+"/"+str(year)

    # convert monthStr to the month name
    months = ["January", "February", "March", "April", 
              "May", "June", "July", "August", 
              "September", "October", "November", "December"]
    monthStr = months[month-1]
    date2 = monthStr+" " + str(day) + ", " + str(year)

    # Output result in month day, year format
    print("The date is {0} or {1}.".format(date1, date2))

main()

Enter the day number: 01
Enter the month number: 01
Enter the year: 2020
The date is 1/1/2020 or January 1, 2020.


2. A certain CS professor give 5-point quizzes that are graded on the scale 5-A, 4-B, 3-C, 2-B, 1-A, 0-F. Write a program that accepts a quiz score as an input and prints out the corresponding grade.

In [ ]:
# Gradeconvert.py 

    # Input a quiz score ranging from 0 - 5
    # Output a grade from A - F
    # Create a sequence of grades as a string, where index/position correspons to letter grade
    # For a number supplied by user, convery to int and get the index, print the appropriate grade

def main()
